src/Workspaces/Core/Portable/Editing/SyntaxEditor.cs

```diff
@@ -36,6 +38,30 @@ internal SyntaxEditor(SyntaxNode root, SyntaxGenerator generator)
             _changes = new List<Change>();
         }
 
+        private SyntaxNode ApplyTrackingToNewNode(SyntaxNode node)
+        {
+            if (node == null)
+            {
+                return null;
+            }
+
+            _lazyTrackedNewNodesOpt = _lazyTrackedNewNodesOpt ?? new HashSet<SyntaxNode>();
+            foreach (var descendant in node.DescendantNodesAndSelf())
```

Comment:
Can be bail on diving into the subtree if the node is already in the hashset? i.e. node.DescendantNodesAndSelf(descendIntoChildren: _lazyTrackedNewNodesOpt.Add(descendant))

Reply:
That was my original implementation. That causes added ReplaceNode unit test to fail if there are subsequent replacements to some descendant node followed by replacement to new node root, i.e. in the example I cited [above](https://github.com/dotnet/roslyn/pull/31558#discussion_r239920445) where that approach fails.